### PR TITLE
credentials/alts: Added error logging for isRunningOnGCP

### PR DESCRIPTION
**Pull request by wflms20110333**
@ZhenLian 
Currently, `isRunningOnGCP` doesn’t return any errors. However, there are reasons that it will fail, which causes `log.Fatal`. This PR enables the function to return error, avoiding any `log.Fatal` calls and changes everything along the call path of this function.

### CODE REVIEW

File: credentials/google/google.go [hunk 1]
```diff
@@ -112,7 +112,11 @@ func (c *creds) NewWithMode(mode string) (credentials.Bundle, error) {
 	case internal.CredsBundleModeBackendFromBalancer, internal.CredsBundleModeBalancer:
 		// Only the clients can use google default credentials, so we only need
 		// to create new ALTS client creds here.
-		newCreds.transportCreds = alts.NewClientCreds(alts.DefaultClientOptions())
+		creds, err := alts.NewClientCreds(alts.DefaultClientOptions())
+		newCreds.transportCreds = creds
+		if err != nil {
+			return nil, fmt.Errorf("failed to create new client credentials: %v", err)
```
Comment (ZhenLian) [inline credentials/google/google.go:118]: Check error before setting creds.

File: interop/alts/client/client.go [hunk 2]
```diff
@@ -42,7 +42,10 @@ func main() {
 	if *hsAddr != "" {
 		opts.HandshakerServiceAddress = *hsAddr
 	}
-	altsTC := alts.NewClientCreds(opts)
+	altsTC, err := alts.NewClientCreds(opts)
+	if err != nil {
+		grpclog.Fatalf("Failed to create new client credentials: %v", err)
```
Comment (ZhenLian) [inline interop/alts/client/client.go:47]: s/Failed/failed

File: interop/alts/server/server.go [hunk 3]
```diff
@@ -60,7 +60,10 @@ func main() {
 	if *hsAddr != "" {
 		opts.HandshakerServiceAddress = *hsAddr
 	}
-	altsTC := alts.NewServerCreds(opts)
+	altsTC, err := alts.NewServerCreds(opts)
+	if err != nil {
+		grpclog.Fatalf("Failed to create new server credentials: %v", err)
```
Comment (ZhenLian) [inline interop/alts/server/server.go:65]: s/Failed/failed/g

**Comment by ZhenLian** [comment]
Thanks @wflms20110333 for the implementation! @cesarghali @easwars might also want to take a look at this PR.

File: credentials/alts/utils.go [hunk 4]
```diff
@@ -83,28 +82,27 @@ var (
...
+		return name == "Google" || name == "Google Compute Engine", nil
 	case "windows":
 		name = strings.Replace(name, " ", "", -1)
 		name = strings.Replace(name, "\n", "", -1)
 		name = strings.Replace(name, "\r", "", -1)
-		return name == "Google"
+		return name == "Google", nil
 	default:
-		log.Fatal(platformError(runningOS))
+		return false, platformError(runningOS)
```
[the remark this answers is not in the document]
Comment (wflms20110333) [reply credentials/alts/utils.go:104]: Should we still log the error then? And I'm guessing that wouldn't be 1`log.Fataf` anymore, since we don't want the code to halt?

[the remark this answers is not in the document]
Comment (wflms20110333) [reply credentials/alts/utils.go:104]: Re: no need to log the error

File: credentials/alts/utils.go [hunk 5]
```diff
@@ -83,28 +82,27 @@ var (
...
+		return name == "Google" || name == "Google Compute Engine", nil
 	case "windows":
 		name = strings.Replace(name, " ", "", -1)
 		name = strings.Replace(name, "\n", "", -1)
 		name = strings.Replace(name, "\r", "", -1)
-		return name == "Google"
+		return name == "Google", nil
 	default:
-		log.Fatal(platformError(runningOS))
+		return false, nil
```
Comment (cesarghali) [inline credentials/alts/utils.go:104]: Should we return `platformError(runningOS)` instead of `nil`?

File: examples/features/encryption/ALTS/server/main.go [hunk 6]
```diff
@@ -50,14 +50,17 @@ func main() {
...
+		log.Fatalf("failed to create new server credentials: %v", err)
+	}
 
 	s := grpc.NewServer(grpc.Creds(altsTC))
 
 	// Register EchoServer on the server.
 	pb.RegisterEchoServer(s, &ecServer{})
 
-	if err := s.Serve(lis); err != nil {
+	if err = s.Serve(lis); err != nil {
```
Comment (cesarghali) [inline examples/features/encryption/ALTS/server/main.go:63]: You can leave this as `:=` to keep the `err` variable local in the `if` :)

File: credentials/alts/utils.go (hunk 5 above)
Comment (ZhenLian) [reply credentials/alts/utils.go:104]: If we return an error, these errors will still be captured by tests. If MacOS users run the tests, I think the tests would still fail? Are there any ways that we can solve this issue(since our original goal is to let MacOS users to be able to run tests in OSS)?

Comment (cesarghali) [reply credentials/alts/utils.go:104]: Ah I see, that's a good point. Then not returning an error is fine with me.

**Comment by easwars** [comment]
Is this ready for merging? Or does it require more reviews?

**Comment by ZhenLian** [comment]
> Is this ready for merging? Or does it require more reviews?

Not yet, this is also used internally so I also have to do some internal cleanup. I will get it done shortly.

File: credentials/alts/alts.go [hunk 7]
```diff
@@ -137,19 +137,24 @@ type altsTC struct {
...
 // NewServerCreds constructs a server-side ALTS TransportCredentials object.
-func NewServerCreds(opts *ServerOptions) credentials.TransportCredentials {
+func NewServerCreds(opts *ServerOptions) (credentials.TransportCredentials, error) {
 	return newALTS(core.ServerSide, nil, opts.HandshakerServiceAddress)
 }
 
-func newALTS(side core.Side, accounts []string, hsAddress string) credentials.TransportCredentials {
+func newALTS(side core.Side, accounts []string, hsAddress string) (credentials.TransportCredentials, error) {
+	var err error
+	err = nil
```
Comment (easwars) [inline credentials/alts/alts.go:151]: `nil` is the zero value for `error`. Why is this line required?

**Comment by ZhenLian** [comment]
Closed in favor of https://github.com/grpc/grpc-go/pull/3904.

File: credentials/alts/alts.go (hunk 7 above)
Comment (ZhenLian) [reply credentials/alts/alts.go:151]: Yeah I think it's not necessary. Fixed in a new PR: https://github.com/grpc/grpc-go/pull/3904. We can continue our discussion there.
